### PR TITLE
Issue #8: calculate number of weeks in year.

### PR DESCRIPTION
**Pull request by babruix**

### CODE REVIEW

**Comment by franleplant** [comment]
WOW! Hi @babruix I wasn't expecting any contributions so early on in this unknown library! I thank you very much for this! Want to give me some context? Are you using this library?
Again, thanks a lot! Feel free to write to my email if you want to chat about things unrelated to this PR but related to the Lib.

Super thanks for this contribution, I would love some changes though:

- Tests: I would like to have test for both `getWeekNumber` and `weeksInYear`. Just export them and test them inside a file called `src/officialValidators/week.test.js`. We are using jest so it should be pretty easily done. Don't need to be super extensive here, just write a couple of `expects` with different parameters, like `2016` and `2015`.
- Format and style: Sorry about not having official docs about this but we are very much WIP. Please replace `var` to either `const` if the value does not change or to `let` if it's going to change.

Thanks, again, a whole lot.
Fran

File: src/officialValidators/week.js [hunk 1]
```diff
@@ -3,20 +3,41 @@
...
-    return true
+    return true;
   }
 
-  return !(0 < year && 1 <= week && week <= MAX_WEEK)
+  return !(0 < year && 1 <= week && week <= weeksInYear(year));
+}
+
+function weeksInYear(year) {
+  var d = new Date(year, 11, 31);
```
Comment (franleplant) [inline src/officialValidators/week.js]: I would go to something like `let date` in here. Or even better `let lastDayOfTheYear`.

File: src/officialValidators/week.js [hunk 2]
```diff
@@ -3,20 +3,41 @@
...
+  var week = getWeekNumber(d)[1];
```
Comment (franleplant) [inline src/officialValidators/week.js]: `var` should be replaced with `const`

File: src/officialValidators/week.js [hunk 3]
```diff
@@ -3,20 +3,41 @@
...
+  
+  return week == 1? getWeekNumber(d.setDate(24))[1] : week;
+}
+
+function getWeekNumber(d) {
+    // Copy date so don't modify original
+    d = new Date(+d);
```
Comment (franleplant) [inline src/officialValidators/week.js]: I would probably copy `d` into a new variable called `let date`.

File: src/officialValidators/week.js [hunk 4]
```diff
@@ -3,20 +3,41 @@
...
+    d.setHours(0,0,0);
+    // Set to nearest Thursday: current date + 4 - current day number
+    // Make Sunday's day number 7
+    d.setDate(d.getDate() + 4 - (d.getDay()||7));
+    // Get first day of year
+    var yearStart = new Date(d.getFullYear(),0,1);
```
Comment (franleplant) [inline src/officialValidators/week.js]: `var` should be replaced with `const`

File: src/officialValidators/week.js [hunk 5]
```diff
@@ -3,20 +3,41 @@
...
+    // Calculate full weeks to nearest Thursday
+    var weekNo = Math.ceil(( ( (d - yearStart) / 86400000) + 1)/7)
```
Comment (franleplant) [inline src/officialValidators/week.js]: `var` should be replaced with `const`

File: src/officialValidators/week.js [hunk 6]
```diff
@@ -3,20 +3,41 @@
...
 
-  return !(0 < year && 1 <= week && week <= MAX_WEEK)
+  return !(0 < year && 1 <= week && week <= weeksInYear(year));
+}
+
+function weeksInYear(year) {
+  var d = new Date(year, 11, 31);
+  var week = getWeekNumber(d)[1];
+  
+  return week == 1? getWeekNumber(d.setDate(24))[1] : week;
```
Comment (franleplant) [inline src/officialValidators/week.js]: `==` should be `===`, right? Are we doing any implicit casting that I might be missing?

File: src/officialValidators/week.js (hunk 2 above)
Comment (franleplant) [inline src/officialValidators/week.js]: I would also use destructuring for better readability

```javascript
const [_, week] = getWeekNumber(d);
```

File: src/officialValidators/week.js (hunk 6 above)
Comment (franleplant) [inline src/officialValidators/week.js]: This chunk could benefit from more clarity and verbosity `getWeekNumber(d.setDate(24))[1]`.
What does it means? Can we assign it to a variable with a proper name that let's us infer the meaning?

**Comment by babruix** [comment]
Thanks for positive feedback and verbose comments 👍 
I do not use library, just thought I would be nice to contribute to well described issue that needs help :)
I will close this PR and create new one, refactored and with tests!

**Pull request closed without merging**